Add channel-reduced and runtime-compressed image formats, rework format detection

illwieckz opened the pull request:
- tr_image: rework the alpha detection
- renderer: implement and detect RED and RG images
- renderer: introduce and use IF_NOALPHA
- tr_backend: fix a comment
- tr_image: flag glyph image with IF_ALPHA
- sdl_glimp: fix the EXT/ARB texture_rg confusion
- tr_image: document the GL_RGB format
- tr_image: add some GL format checks and asserts
- tr_image: reshape the cinematic image generation a bit
- tr_image: reshape the default image painting a bit
- tr_image: make some iterators more local

---

Previous message:

- tr_image: reword the fog generation code
That's just beautifying, following patches are written above this
- tr_image: store the fog alpha channel in red one
Imported from #1730
but without the colorspace-related HACKs
- renderer: add support for the GL_RED image format
Imported from #1730
- tr_image: upload _fog image as GL_RED
Imported from #1730
- tr_image: upload _black, _red and _blackCube images as GL_RED
- tr_image: rework the alpha detection
- tr_image: detect RED images
- renderer: introduce and use IF_NOALPHA
Useful for when we know in advance an image has no alpha channel
- tr_image: do not rely on previously set data for _cinematic, make sure it cannot be detected as RED
- tr_backend: fix a comment
Actually `GL_COMPRESSED_*` formats can be used (they let the driver decide), but we don't use them.

illwieckz committed:
renderer: implement and detect RED and RG images

illwieckz committed:
tr_image: rework the alpha detection

illwieckz committed:
renderer: introduce and use IF_NOALPHA

illwieckz committed:
tr_backend: fix a comment

slipher commented:
> Oh, that's interesting! Would you also know what computation to expect when blending the linear way? Or maybe it should have been `sqrt()` in linear space but the naive blending did garbage and we're accustomed to it?

Square root doesn't make much sense as a model for the fog; probably they just tried random functions and picked something that looked OK. Instead of `sqrt(a*x)`, `1 - exp(b*x)` would be an obvious model to use (as if the fog is formed from layered alpha blending). `a`/`b` is a constant based on the fog density. You can make these functions match up somewhat using appropriate constants: https://www.desmos.com/calculator/abvkogdqzy

illwieckz committed:
tr_image: flag glyph image with IF_ALPHA

illwieckz commented:
That looks ready to me.

illwieckz commented:
I fixed an non-optimal code path, it was testing for RED before testing for ALPHA, meaning a red texture with an opaque alpha channel was only detected as RGB, not as RED. Now fixed:

```
]/listImages *red*
------------------------------------------------------------------------- 
num   width height layers mm  type format   space  wrap.t   wrap.s   name 
------------------------------------------------------------------------- 
3         8      8      1 no  2D   R8       linear t.rept   s.rept   _red 
99       64     64      1 no  2D   DXT5     linear t.clmp   s.clmp   emoticons/featured 
119     256    256      1 yes 2D   RGB8     linear t.rept   s.rept   textures/pulse/metal-red.tga 
120     256    256      1 yes 2D   RGB8     linear t.rept   s.rept   textures/pulse/metal-red 
158     256    256      1 yes 2D   RGB8     linear t.rept   s.rept   textures/pulse/e8base_red 
159     256    256      1 yes 2D   RGB8     linear t.rept   s.rept   textures/pulse/e8base_red.tga 
233      16     16      1 no  2D   R8       linear t.rept   s.rept   textures/pulse/red16x16.tga 
262      16     16      1 yes 2D   DXT1     linear t.rept   s.rept   gfx/buildables/common/redbuild 
281      16    256      1 yes 2D   RGB8     linear t.rept   s.rept   gfx/cgrading/red-only 
------------------------------------------------------------------------- 
0 total texels (not including mipmaps) 
1.01 MB total image memory (estimated) 
689 total images 
------------------------------------------------------------------------- 
```

Unrelated to that branch, but I notice we have both `textures/pulse/metal-red.tga` and `textures/pulse/metal-red` loaded… that doesn't look right.

illwieckz committed:
renderer: make possible to compress uncompressed textures

Add:

- r_compressColormaps
- r_compressSkyboxes

illwieckz committed:
tr_image: document the GL_RGB format

illwieckz committed:
tr_image: add some GL format checks and asserts

illwieckz commented:
This now also support and detect RG textures, and use explicit `GL_R8` instead of `GL_RED` when supported.

```
]/listImages *16x16*
-------------------------------------------------------------------------
num   width height layers mm  type format   space  wrap.t   wrap.s   name
-------------------------------------------------------------------------
205      16     16      1 no  2D   R8       linear t.rept   s.rept   textures/pulse/black16x16.tga
280      16     16      1 yes 2D   RGB8     linear t.rept   s.rept   textures/pulse/white16x16.tga
283      16     16      1 no  2D   RGB8     linear t.rept   s.rept   textures/pulse/white16x16.tga
291      16     16      1 yes 2D   R8       linear t.rept   s.rept   textures/pulse/black16x16.tga
296      16     16      1 no  2D   R8       linear t.rept   s.rept   textures/pulse/red16x16.tga
302      16     16      1 no  2D   RG8      linear t.rept   s.rept   textures/pulse/green16x16.tga
-------------------------------------------------------------------------
0 total texels (not including mipmaps)
0.00 MB total image memory (estimated)
752 total images
-------------------------------------------------------------------------
```

illwieckz commented:
> This now also support and detect RG textures, and use explicit `GL_R8` instead of `GL_RED` when supported.

Actually implementing RG support fixed a dormant bug where `glConfig.textureRGAvailable` was never set to true but we had code testing for it.

illwieckz commented:
So yes there was a pre-existing confusion on `master` related to some RG extension but it was less problematic than what I thought at first, it was just a naming mess.

illwieckz commented:
I found a bug in the code in `master` where we only iterate the first layer of an image for detecting the format. Now fixed.

illwieckz commented:
I noticed that would modify an undocumented feature of `_black`, previously it was usable to make an transparent texture:

```
transparent
{
 map _black
 blend blend
}
```

Because all its pixels were set to 0, but when stored as RED the alpha channel is opaque (so, 1).

But I have seen no map in the whole betaserv corpus that used any `_black`, `*black`, `$black`, `$blackimage` keywords at all.

Actually it bugged me that the black image was not opaque black.

illwieckz committed:
tr_image: reshape the cinematic image generation a bit

illwieckz committed:
tr_image: reshape the default image painting a bit

illwieckz committed:
tr_image: make some iterators more local

illwieckz commented:
Detecting the RG images is less needed after we all the color variants, as seen in:

- https://github.com/DaemonEngine/Daemon/pull/1804
- https://github.com/UnvanquishedAssets/tex-common_src.dpkdir/pull/9

Though, this can still be useful for optimizing the upload of legacy images from legacy paks, that are not compressed, and then for which we can detect if they use their blue channel or not.

This PR actually fixes a bug in detecting the alpha channel presence (it only looked for the first layer of an image).

This also adds `IF_NOALPHA` to skip the detection of the alpha channel when it is known it is fully opaque.

Something not implemented in that PR is to flag images with `IF_NOALPHA` when they are part of a shader without blend functions, something we can do, but I'm postponing this. Especially, we may need some revamp to be done first, as some images are loaded before the stage is entirely parsed (and then the blending is unknown). But that's likely something I want to do later.

Also something I want to do later is to reimplement the GL compression of non-compressed images (in the meaning of non-DXT compressed images), for use with the legacy assets full of non-crunched JPG or TGA files. I actually have a working branch for that based on an older version of this branch, but then again that's for later.

slipher reviewed (commented):
How much memory do we expect to save from this red image detection?

illwieckz commented:
> How much memory do we expect to save from this red image detection?

That's one of my concern. Now that we ditched the color images, and after we will use q3shader code for common color images instead of pixmaps, not only RG detection but also RED detection becomes less useful.

This can still benefit some legacy maps, for example the `pulse` maps has some plain red and green images for displaying the door closed/open status in the control room, legacy maps without may have things like that, pure black, red or green images, the win isn't very big.

illwieckz commented on src/engine/renderer/tr_image.cpp:
```diff
@@ -2794,12 +2794,14 @@ void R_CreateBuiltinImages()
+	/* Generate cinematic frames.
+	It is empty data to be filled by the cinematic code, but
+	we fill it with non-zero values so the format detector keeps all color channels. */
+	memset( data, 255, sizeof( data ) );
```
I have no idea, but that would have to be investigated for another PR, not this one.

illwieckz commented:
I rebased this. I'm not sure everything has to be merged, but some patches here are fixing real bugs, and have to be merged, separately or not.

I also imported other patches that was sitting in another branch based on this one, that I expected to submit after this one would have been merged.

The motivation for that work was to please @sweet235 as he expressed the concern of uncompressed images taking a lot of VRAM. It's very common when mapping to run the work-in-progress map with the uncompressed textures, before doing a nice dpk with crn, etc.

The bug fixes were found along the way while modifying the code I was expanding.

Those other patches I imported are to reimplement texture compression for when textures aren't crunched in dpk. This can also be useful for legacy maps.

So this texture compression feature and the other texture detection (especially to detect grayscale images and load them as single-channel R8 images) were purposed to make mapping easier by reducing the VRAM usage when loading work-in-progress maps.

The compression is cheap, fast and then, not as good as crunch, but that looks good enough for textures. So my idea was that compression for uncompressed colormaps would be enabled in high, medium, low and lowest presets, keeping uncompressed images uncompressed in ultra preset.

The compression of skyboxes is a separate option, because a skybox has the particularity to not move on screen so less good textures are more obvious. That's why to begin with we ship skyboxes as lossy webp (which is good compression with almost unnoticeable artifacts even if not lossless), but then we can decide to keep them as is in high and ultra preset, while compressing them in medium and low.

Mappers would be able to enable compression on both colormaps and skyboxes whatever their base preset just to reduce VRAM usage when loading work-in-progress maps.

The texture compression feature isn't meant to encourage people to not crunch their textures, but to reduce VRAM usage when loading quickly repackaged or not-repackaged legacy maps, or when mapping, before packaging them.

With official assets, both R8 and colormap compression would never be used because Urcheon and Crunch do all the magic at packaging time, but the skybox compression is allowing us to ship nice looking skyboxes for high preset while compressing them in low presets for performance.

slipher commented:
> The motivation for that work was to please @sweet235 as he expressed the concern of uncompressed images taking a lot of VRAM. It's very common when mapping to run the work-in-progress map with the uncompressed textures, before doing a nice dpk with crn, etc.

How does that use case relate to this PR about adding GL_RED? Is Sweet making a monochrome map with dozens of textures that are all completely red?

illwieckz commented:
> Is Sweet making a monochrome map with dozens of textures that are all completely red?

Lol no, it was just part of the whole panorama of optimized formats I investigated.

The motivation was to implement more optimized formats, and some we could select at run time, so I looked at what was available.

There is no direct link with mapping and red images, but there are direct links with mapping and optimized format, and direct link with optimized formats and channel-specific storage.

illwieckz commented:
For example in the future we may load uncompressed normal maps as RG (XY) since we reconstruct Z.
We may also load uncompressed height maps as R and sample from R in such case, etc.

illwieckz commented:
We may also load in the future uncompressed ORM maps (for PBR) as RM in RG with a swizzle. It's actually sad that the common standard is ORM while O is the most optional thing of them (at least that's the one we don't implement yet). 😂️

illwieckz commented:
Speaking of swizzles, I also have plans for implementing some texture swizzles when available to avoid some shader permutations and other things. It's all part of a bigger picture.
